packages/astro/src/content/vite-plugin-content-imports.ts

```diff
@@ -41,116 +45,201 @@ export function astroContentImportPlugin({
+	const contentEntryModuleByIdCache = new Map<string, ContentEntryModule | 'loading'>();
+	const awaitingCacheById = new Map<string, ((val: ContentEntryModule) => void)[]>();
+	function getContentEntryModuleFromCache(id: string) {
+		const value = contentEntryModuleByIdCache.get(id);
+		// It's possible for Vite to load modules that depend on this cache
+		// before the cache is populated. In that case, we queue a promise
+		// to be resolved by `setContentEntryModuleCache`.
+		if (value === 'loading') {
+			return new Promise<ContentEntryModule>((resolve, reject) => {
+				const awaiting = awaitingCacheById.get(id) ?? [];
+				awaiting.push(resolve);
+				awaitingCacheById.set(id, awaiting);
+			});
+		} else if (value) {
+			return Promise.resolve(value);
+		}
+		return Promise.resolve(undefined);
+	}
+
+	async function setContentEntryModuleCache({
+		fileId,
+		pluginContext,
+	}: {
+		fileId: string;
+		pluginContext: PluginContext;
+	}): Promise<ContentEntryModule> {
+		contentEntryModuleByIdCache.set(fileId, 'loading');
+		const observable = globalContentConfigObserver.get();
+
+		// Content config should be loaded before this plugin is used
+		if (observable.status === 'init') {
+			throw new AstroError({
+				...AstroErrorData.UnknownContentCollectionError,
+				message: 'Content config failed to load.',
+			});
+		}
+		if (observable.status === 'error') {
+			// Throw here to bubble content config errors
+			// to the error overlay in development
+			throw observable.error;
+		}
+
+		let contentConfig: ContentConfig | undefined =
+			observable.status === 'loaded' ? observable.config : undefined;
+		if (observable.status === 'loading') {
+			// Wait for config to load
+			contentConfig = await new Promise((resolve) => {
+				const unsubscribe = globalContentConfigObserver.subscribe((ctx) => {
+					if (ctx.status === 'loaded') {
+						resolve(ctx.config);
+						unsubscribe();
+					} else if (ctx.status === 'error') {
+						resolve(undefined);
+						unsubscribe();
+					}
+				});
+			});
+		}
+		const rawContents = await fs.promises.readFile(fileId, 'utf-8');
+		const fileExt = extname(fileId);
+		if (!contentEntryExtToParser.has(fileExt)) {
+			throw new AstroError({
+				...AstroErrorData.UnknownContentCollectionError,
+				message: `No parser found for content entry ${JSON.stringify(
+					fileId
+				)}. Did you apply an integration for this file type?`,
+			});
+		}
+		const contentEntryParser = contentEntryExtToParser.get(fileExt)!;
+		const info = await contentEntryParser.getEntryInfo({
+			fileUrl: pathToFileURL(fileId),
+			contents: rawContents,
+		});
+		const generatedInfo = getEntryInfo({
+			entry: pathToFileURL(fileId),
+			contentDir: contentPaths.contentDir,
+		});
+		if (generatedInfo instanceof NoCollectionError) throw generatedInfo;
+
+		const _internal = { filePath: fileId, rawData: info.rawData };
+		// TODO: move slug calculation to the start of the build
```

Comment:
Is there an issue for tracking this TODO?

Reply:
Nope, but there should be!